engine/src/main/resources/assets/shaders/chunk_frag.glsl

```diff
@@ -92,34 +92,41 @@ void main() {
     vec3 normalOpaque = normal;
     float shininess = 0.0;
 
-#if defined (NORMAL_MAPPING)
+#if defined (NORMAL_MAPPING) || defined (PARALLAX_MAPPING)
     // TODO: Calculates the tangent frame on the fly - this is absurdly costly... But storing
     // the tangent for each vertex in the chunk VBO might be not the best idea either.
-    vec3 dp1 = dFdx(vertexProjPos.xyz);
-    vec3 dp2 = dFdy(vertexProjPos.xyz);
-    vec2 duv1 = dFdx(gl_TexCoord[0].xy);
-    vec2 duv2 = dFdy(gl_TexCoord[0].xy);
-
-    vec3 dp2perp = cross(dp2, normal);
-    vec3 dp1perp = cross(normal, dp1);
-    vec3 tangent = dp2perp * duv1.x + dp1perp * duv2.x;
-    vec3 binormal = dp2perp * duv1.y + dp1perp * duv2.y;
-
-    float invMax = inversesqrt(max(dot(tangent,tangent), dot(binormal,binormal)));
-    mat3 tbn = mat3(tangent * invMax, binormal * invMax, normal);
+    // Screen-space coordinates happen to be a coordinate system to which it's possible to relate both view-space and texture coordinates, but are themselves irrelevant.
+    mat2x3 screenToView = mat2x3(dFdx(vertexViewPos.xyz), dFdy(vertexViewPos.xyz));
+    mat2   screenToUv   = mat2  (dFdx(gl_TexCoord[0].xy), dFdy(gl_TexCoord[0].xy)) / TEXTURE_OFFSET;
```

Comment:
These two variables store matrices to transform from one space to another: I'd rename them accordingly: `screenSpaceToViewSpace` and `screenSpaceToUvSpace`. I'd also rename vertexViewPos to the less ambiguous vertexPositionInViewSpace - assuming that's what it means.

Reply:
I suppose I've referred to them like that myself in the comments too, but it's probably better to call them change-of-basis matrices than referring to them as converting between different spaces. In this case though some of the coordinate systems used aren't actually bases (or are only bases for the tangent space to the mesh, not to view space), as they have too few elements.

Reply:
In this case these are mostly just different bases for the same space (or a subspace of it for the 2D ones), as there are natural maps between these coordinate systems that make it seem sensible to consider them as different representations of the same underlying vector space.